tests/disp_dev/main.c

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright (C) 2020 Inria
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup tests
+ * @{
+ *
+ * @file
+ * @brief       Generic display device test application
+ *
+ * @author      Alexandre Abadie <alexandre.abadie@inria.fr>
+ *
+ * @}
+ */
+
+#include <stdio.h>
+#include <assert.h>
+
+#include "disp_dev.h"
+
+#include "ili9341.h"
+#include "ili9341_params.h"
+#include "ili9341_disp_dev.h"
+
+#include "riot_logo.h"
+
+#include "test_utils/expect.h"
+
+static ili9341_t ili9341;
+
+int main(void)
+{
+#ifdef BOARD_PINETIME
+    /* on PineTime, enable the backlight */
+    gpio_clear(LCD_BACKLIGHT_LOW);
+#endif
+
+    ili9341_init(&ili9341, &ili9341_params[0]);
+
+    disp_dev_t *dev = (disp_dev_t *)&ili9341;
+    dev->driver = &ili9341_disp_dev_driver;
+
+    disp_dev_set_invert(dev, true);
+
+    uint16_t max_width = disp_dev_width(dev);
+    uint16_t max_height = disp_dev_height(dev);
+
+    expect(max_width == ili9341.params->lines);
+    expect(max_height == 240);
+
+    uint16_t color = 0;
+    for (uint16_t x = 0; x < max_width; ++x) {
+        for (uint16_t y = 0; y < max_height; ++y) {
```

Comment:
Seeing this is a bit painful.

It screams for a "fill(x,x2,y,y2, color)" that doesn't do a `map()` for every pixel. But we can leave that to a followup.

Iterating linewise (first all x=0 y=0, 1, 2, ... then all x=1, y=0, 1, 2) is also a textbook example for missing optimization. Please change to `for (y, ...) for (x...)`.

Reply:
> Please change to for (y, ...) for (x...)

Done but doesn't make much difference in speed.

Reply:
> It screams for a "fill(x,x2,y,y2, color)"

Indeed, we also had this discussion with @bergzand IRL. We came to the conclusion that `map` can also be added in this situation with an input buffer. The input doesn't necessarily need to have a size equal to the size of the screen but then that would require rectangle size computation and multiple calls to map in users code (so not very handy, but more generic).

Reply:
The only way to speed this up is to use a buffer with the pixels for the selected color and transmit those. Even a buffer of only 32 pixels (64 bytes) already speeds this up tremendously.

This because (at least for the ili9341), each `map()` first sends `x1, x2, y1, y2` (8 bytes) to the display and then the pixels (another 2 byte). That's 10 bytes of SPI data for a single transfer. Transmitting a buffer of 32 pixels (64 byte + 8 byte coordinates = 72 bytes) at once is a lot more efficient in terms of SPI transfers.